test/unit/master/test_build.py

```diff
@@ -17,230 +18,178 @@
-            build.complete_subjob(subjob.subjob_id(), payload=payload)
+        build = self._create_test_build(BuildStatus.BUILDING)
+        subjob = build.all_subjobs()[0]
 
-    def test_need_more_slaves_returns_false_if_max_processes_is_reached(self):
-        subjobs = self._create_subjobs(count=5)
-        mock_project_type = self._create_mock_project_type()
-        mock_slave = self._create_mock_slave(num_executors=1)
-        build = Build(BuildRequest({}))
-        build._project_type = mock_project_type
+        self.mock_util.fs.write_file.side_effect = FileExistsError
 
-        build.prepare(subjobs, self._create_job_config(max_executors=1))
-        build.allocate_slave(mock_slave)
-        self.assertFalse(build.needs_more_slaves(), "if max processes is reached, we shouldn't need more slaves")
+        with self.assertRaises(Exception):
+            build.complete_subjob(subjob.subjob_id(), payload=self._FAKE_PAYLOAD)
+
+    @genty_dataset(
+        max_executors_reached=(1, False),
+        max_executors_not_reached=(30, True),
+    )
+    def test_need_more_slaves_returns_false_if_and_only_if_max_executors_is_reached(
+            self,
+            max_executors_for_build,
+            build_should_need_more_slaves
+    ):
+        job_config = self._create_job_config(max_executors=max_executors_for_build)
+        build = self._create_test_build(BuildStatus.PREPARED, num_subjobs=100, job_config=job_config)
 
-    def test_need_more_slaves_returns_true_if_max_processes_is_not_reached(self):
-        subjobs = self._create_subjobs(count=8)
-        mock_project_type = self._create_mock_project_type()
         mock_slave = self._create_mock_slave(num_executors=5)
-        build = Build(BuildRequest({}))
-        build._project_type = mock_project_type
+        build.allocate_slave(slave=mock_slave)
 
-        build.prepare(subjobs, self._create_job_config(max_executors=8))
-        build.allocate_slave(mock_slave)
-        self.assertTrue(build.needs_more_slaves(), "if max_processes is not reached, we should need more slaves")
+        self.assertEqual(build.needs_more_slaves(), build_should_need_more_slaves,
+                         'If and only if the maximum number of executors is allocated we should not need more slaves.')
 
     def test_build_cannot_be_prepared_more_than_once(self):
-        build = Build(BuildRequest({}))
-        subjobs = self._create_subjobs(count=3)
-        mock_project_type = self._create_mock_project_type()
-        build._project_type = mock_project_type
+        build = self._create_test_build(BuildStatus.QUEUED)
+        job_config = self._create_job_config()
+        subjobs = self._create_subjobs(count=3, job_config=job_config)
 
-        build.prepare(subjobs, self._create_job_config())
+        build.prepare(subjobs=subjobs, job_config=job_config)
 
-        with self.assertRaises(Exception):
-            build.prepare(subjobs, mock_project_type, self._create_job_config())
```

Comment:
I think this was actually not testing this correctly before -- the exception that was being raised was around the extra argument to `prepare()`, not the one we were trying to test for.